lib/galaxy/files/sources/__init__.py

```diff
@@ -405,18 +432,33 @@ def list(
         recursive=False,
         user_context: "OptionalUserContext" = None,
         opts: Optional[FilesSourceOptions] = None,
-    ) -> List[AnyRemoteEntry]:
+        limit: Optional[int] = None,
+        offset: Optional[int] = None,
+        query: Optional[str] = None,
+        sort_by: Optional[str] = None,
+    ) -> Tuple[List[AnyRemoteEntry], int]:
         self._check_user_access(user_context)
-        return self._list(path, recursive, user_context, opts)
+        if not self.supports_pagination and (limit is not None or offset is not None):
+            raise RequestParameterInvalidException("Pagination is not supported by this file source.")
+        if not self.supports_search and query:
+            raise RequestParameterInvalidException("Server-side search is not supported by this file source.")
+        if not self.supports_sorting and sort_by:
+            raise RequestParameterInvalidException("Server-side sorting is not supported by this file source.")
+
```

Comment:
I think this may be also a good place to verify that both `limit` and `offset` are non-negative. We need that verification because a negative value for either one is meaningless, but also could trigger errors (one example of that would be a division by zero in `page = (offset or 0) // size + 1` if `limit == -1`)